### Using Textblob for NLP/Sentiment Analysis 

In [99]:
import pandas as pd
from textblob import TextBlob, Word, Blobber 
import matplotlib.pyplot as plt
% matplotlib inline
from nltk.corpus import stopwords

TextBlob returns polarity and subjectivity scores for the text is takes in:

In [100]:
TextBlob("Everything is awesome!").sentiment

Sentiment(polarity=1.0, subjectivity=1.0)

### **Polarity**: 

***Polarity***, is the emotion expressed in the sentence - it addresses the tone of the text. It can be positive, negative or neutral.

- #polarity (-1 to 1): scores closer to -1 are more negative in tone, closer to 0 are more neutral, and closer to 1 are more positive in tone.

### **Subjectivity**:

**Subjectivity**: Subjective sentence expresses some personal feelings, views, or beliefs.

- #subjectivity (0 to 1): scores closer to 0 are more objective in tone, scores closer to 1 are more subjective in tone



In [101]:
TextBlob("").sentiment.polarity


0.0

In [102]:

TextBlob("").sentiment.subjectivity


0.0

#### Cool things about sentiment analysis with TextBlob:
    - It doesn't care about punctuation
    - It doesn't care about capitalization
#### Drawback:
    - It just ignores words it doesn't understand and gives them a score of 0.0

In [103]:
TextBlob('bodacious, dude').sentiment

Sentiment(polarity=0.0, subjectivity=0.0)

#### Other cool things it can do:

In [104]:
text = "An artificial man, who was incompletely constructed and has scissors for hands, leads a solitary life. Then one day, a suburban lady meets him and introduces him to her world. "
##this text was taken from imdb: https://www.imdb.com/title/tt0099487/?ref_=nv_sr_1

blob = TextBlob(text)


#### Tags:

TextBlob can also tell us what part of speech each word in a text corresponds to. It can tell us if a word in a sentence is functioning as a noun, an adjective, a verb, etc. In NLP, associating a word with a part of speech is called “tagging.” For a list of what all the different tags mean, check out this website: https://www.clips.uantwerpen.be/pages/mbsp-tags

In [105]:
# Tagging: 

blob.tags

# Examples:
# JJ - adjective
# NN - noun
# RB - adverbs

[('An', 'DT'),
 ('artificial', 'JJ'),
 ('man', 'NN'),
 ('who', 'WP'),
 ('was', 'VBD'),
 ('incompletely', 'RB'),
 ('constructed', 'VBN'),
 ('and', 'CC'),
 ('has', 'VBZ'),
 ('scissors', 'NNS'),
 ('for', 'IN'),
 ('hands', 'NNS'),
 ('leads', 'VBZ'),
 ('a', 'DT'),
 ('solitary', 'JJ'),
 ('life', 'NN'),
 ('Then', 'RB'),
 ('one', 'CD'),
 ('day', 'NN'),
 ('a', 'DT'),
 ('suburban', 'JJ'),
 ('lady', 'NN'),
 ('meets', 'VBZ'),
 ('him', 'PRP'),
 ('and', 'CC'),
 ('introduces', 'VBZ'),
 ('him', 'PRP'),
 ('to', 'TO'),
 ('her', 'PRP$'),
 ('world', 'NN')]

Noun_phrases:

In [106]:
blob.noun_phrases  #combinations of words -  descriptor+ noun

WordList(['artificial man', 'solitary life', 'suburban lady'])

#### Sentence Parsing: 

In [107]:
blob.sentences  #will split the text into individual sentences

[Sentence("An artificial man, who was incompletely constructed and has scissors for hands, leads a solitary life."),
 Sentence("Then one day, a suburban lady meets him and introduces him to her world.")]

In [108]:
for sentence in blob.sentences:  #we can then loop through the sentences to get scores
    print(sentence.sentiment)


Sentiment(polarity=-0.6, subjectivity=1.0)
Sentiment(polarity=0.0, subjectivity=0.0)


#### Can remove Stopwords:

In [109]:
less_words = [word for word in blob.words if word not in stopwords.words('english')]
less = ' '.join(less_words)

new_blob = TextBlob(less)

new_blob


TextBlob("An artificial man incompletely constructed scissors hands leads solitary life Then one day suburban lady meets introduces world")

#### Translation

In [112]:
#Can translate the text into other languages:

blob.translate(to="es")

TextBlob("Un hombre artificial, que fue construido de forma incompleta y tiene tijeras para las manos, lleva una vida solitaria. Entonces, un día, una dama de los suburbios se encuentra con él y lo presenta a su mundo.")

### Hillary Clinton's Presidential Debate Text

### Debate 1 sentiment overall:

In [117]:
with open('ClintonPresDebate1.txt', 'r', encoding = "ISO-8859-1" ) as myfile:
    hillary = myfile.read()

In [121]:
text = TextBlob(hillary)

In [122]:
text.sentiment.subjectivity
#subjectivity (0 to 1): scores closer to 0 are more objective in tone, scores closer to 1 are more subjective in tone

0.4748706812745692

In [123]:
text.sentiment.polarity 
#polarity (-1 to 1): scores closer to -1 are more negative in tone, closer to 0 are more neutral, and closer to 1 are more positive in tone.

0.1679563135448666

 In Debate 1:
-  Clinton Subjectivity: 0.4748706812745692 (slightly more objective than Trump overall)
-  Trump Subjectivity: 0.527053192826821

        
-  Clinton Polarity: 0.1679563135448666 (slightly more positive in tone overall)
-  Trump Polarity:  0.12492547933096725

In [127]:
# text.noun_phrases[:25]

WordList(['donald', 'lester', 'hofstra', 'central question', "granddaughter 's", 'new jobs', 'good jobs', 'renewable energy', 'small business', 'new jobs', 'small business', 'economy fairer', 'national minimum wage', "women 's work", 'support people', 'balance family', 'difficult choices', 'sick days', 'affordable child care', 'debt-free college', 'fair share', 'corporate loopholes', 'donald trump', 'i. donald', 'important issues'])

In [130]:
for sentence in text.sentences:
    print(sentence.sentiment.polarity)

0.0
0.2
0.27999999999999997
0.0
0.375
0.41818181818181815
0.0
0.0
0.15050505050505047
0.0
0.0
0.5
0.5
0.5
0.0
-0.7142857142857143
0.5
0.0
0.39999999999999997
0.35
0.4
0.4
0.5
0.4
-0.125
0.45714285714285713
0.0
0.6
0.2916666666666667
0.25
0.0
0.3333333333333333
0.0
0.0
0.52
0.425
0.0
0.0
0.0
-0.2916666666666667
-0.1027777777777778
0.3333333333333333
0.6
0.0
-0.3
0.24285714285714288
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.43333333333333335
0.05
0.25
0.37878787878787873
0.3666666666666667
0.18333333333333335
0.0
0.2
0.2
0.4
0.0
0.16666666666666669
0.18333333333333335
0.16818181818181818
0.16818181818181818
0.041666666666666664
0.0
-0.05
0.475
0.13636363636363635
0.0
0.0
0.0
0.0
0.7
0.0
0.0
0.0
0.0
0.0
0.35714285714285715
0.0
0.0
0.0
0.2431818181818182
0.0
0.0
-0.20000000000000004
0.0
0.0
0.36818181818181817
0.3
0.0
0.35
0.4
0.3181818181818182
0.0
0.0
0.0
0.1
0.5666666666666667
0.0
0.2
0.0
0.0
0.6
0.0
0.0
-0.25
0.5
0.6
-0.3
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
-0.175
0.5
0.357

In [132]:
# text.translate(to="es")

TextBlob("-ŻCómo estás, Donald?
-Bueno, gracias, Lester, y gracias a Hofstra por hospedarnos. La pregunta central en esta elección es realmente qué tipo de país queremos ser y qué tipo de futuro construiremos juntos. Hoy es el segundo cumpleaños de mi nieta, así que pienso mucho en esto. Primero, tenemos que construir una economía que funcione para todos, no solo para los que están arriba. Eso significa que necesitamos nuevos empleos, buenos empleos, con ingresos crecientes.
-Quiero que invirtamos en ti. Quiero que invirtamos en tu futuro. Eso significa trabajos en infraestructura, manufactura avanzada, innovación y tecnología, energía limpia, renovable y pequeñas empresas, porque la mayoría de los nuevos trabajos provendrán de pequeñas empresas. También tenemos que hacer que la economía sea más justa. Esto comienza con el aumento del salario mínimo nacional y también garantiza, finalmente, un salario igual para el trabajo de las mujeres.
-También quiero ver a más compañías que compart